% Template for PLoS
% Version 1.0 January 2009
%
% To compile to pdf, run:
% latex plos.template
% bibtex plos.template
% latex plos.template
% latex plos.template
% dvipdf plos.template

\documentclass[10pt]{article}

% amsmath package, useful for mathematical formulas
\usepackage{amsmath}
% amssymb package, useful for mathematical symbols
\usepackage{amssymb}

% graphicx package, useful for including eps and pdf graphics
% include graphics with the command \includegraphics
\usepackage{graphicx}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

\usepackage{color} 

% Use doublespacing - comment out for single spacing
%\usepackage{setspace} 
%\doublespacing


% Text layout
\topmargin 0.0cm
\oddsidemargin 0.5cm
\evensidemargin 0.5cm
\textwidth 16cm 
\textheight 21cm

% Bold the 'Figure #' in the caption and separate it with a period
% Captions will be left justified
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

% Use the PLoS provided bibtex style
\bibliographystyle{plos2009}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother


% Leave date blank
\date{}

\pagestyle{myheadings}
%% ** EDIT HERE **


%% ** EDIT HERE **
%% PLEASE INCLUDE ALL MACROS BELOW

%% END MACROS SECTION

\begin{document}

% Title must be 150 characters or less
\begin{flushleft}
{\Large
\textbf{The khmer software package: enabling efficient sequence analysis}
}
% Insert Author names, affiliations and corresponding author email.
\\

Michael R. Crusoe, Greg Edvenson, Jordan Fish, Adina Howe, Luiz Irber, Eric McDonald, Joshua Nahum, Kaben Nanlohy, Humberto Ortiz-Zuazaga, Jason Pell, Jared Simpson, Camille Scott, Ramakrishnan Rajaram Srinivasan, Qingpeng Zhang, and C. Titus Brown$^{\ast}$
\\
$\ast$ E-mail: Corresponding ctb@msu.edu
\end{flushleft}

% Please keep the abstract between 250 and 300 words
\section*{Abstract}

The khmer package is a freely available software library for working
efficiently with fixed length DNA words, or k-mers.  khmer provides
implementations of a probabilistic k-mer counting data structure,
a compressible De Bruijn graph representation, De Bruijn graph
partitioning, and digital normalization.  khmer is implemented in C++
and Python, and is freely available under the BSD license at http://github.com/ged-lab/khmer/.

% Please keep the Author Summary between 150 and 200 words
% Use first person. PLoS ONE authors please skip this step. 
% Author Summary not valid for PLoS ONE submissions.   
%\section*{Author Summary}

\section*{Introduction}

DNA words of a fixed-length k, or ``k-mer'', are a common abstraction
in DNA sequence analysis that enable alignment-free sequence analysis
and comparison. With the advent of second-generation
sequencing and the widespread adoption of De Bruijn graph-based
assemblers, k-mers became even more widely used.  However, the
dramatically increased rate of sequence data generation from the Roche
454 and the Solexa/Illumina sequencers has challenged the basic
data structures and algorithms for k-mer storage and manipulation.
This has led to the development of a wide range of data structures and
algorithms that explore possible improvements to k-mer-based
approaches.

Here we present version 1.0 of the khmer software package, a
high-performance library implementing memory- and time-efficient
algorithms for the manipulation and analysis of short-read data sets.  khmer
contains reference implementations of several approaches, including a
probabilistic k-mer counter based on the CountMin Sketch \cite{Zhang2013}, a
compressible De Bruijn graph representation built on top of Bloom
filters \cite{Pell2012}, and a streaming lossy compression approach for
short-read data sets termed ``digital normalization'' \cite{Brown2012}.

khmer is both research software and a software product for users: it
has been used in the development of novel data structures and
algorithms, and it is also immediately useful for certain kinds of
data analysis (discussed below).  We are continuing to develop research
extensions while maintaining existing functionality.

The khmer software consists of a core library implemented in C++, a
CPython library wrapper implemented in C, and a set of Python
``driver'' scripts that make use of the library to perform various
sequence analysis tasks.  The software is currently developed on
GitHub under github.com/ged-lab/, and it is released under the BSD
License.  There is greater than 80\% statement coverage under
automated tests, measured on both C++ and Python code but primarily
executed at the Python level.

% Results and Discussion can be combined.
\section*{Software details}

\subsection*{Software design}

The design of the khmer package is centered on hashtables, which are
the primary data structure used for both k-mer counting and k-mer
graph storage.  The CountMin Sketch implementation, detailed in \cite{Zhang2013},
uses N hash tables with 8-bit counters; the De Bruijn graph implementation
uses N hash tables with 1-bit counters.  Graph traversal and
graph labeling functionality that support the partitioning algorithm
in \cite{Pell2012} are built on top of the hashtable class in a separate ``subset''
class.

\subsection*{Implementation}

The core data structures and traversal code are implemented in C++, and
then wrapped for Python in hand-written C code, for a total of
12.2k lines of C/C++ code.  The command-line API and all of the tests
are written in 6.6k lines of Python code.

Documentation is written in reStructuredText, compiled with Sphinx, and
hosted on ReadTheDocs.org.

\section*{Features}

khmer has several complementary feature sets, all centered on short-read
manipulation and filtering.

\subsection*{k-mer counting and read trimming}

Using a memory-efficient CountMin Sketch approach, khmer provides an
interface for online counting of k-mers in streams of reads.  The
basic functionality includes calculating the k-mer frequency spectra
in sequence data sets and trimming reads at low-abundance k-mers.
This functionality is explored in \cite{Zhang2013}.

\subsection*{De Bruijn graph representation and exploration}

We have also built a De Bruijn graph representation on top of a Bloom
filter, and implemented this in khmer.  The primary use for this so
far has been to enable memory efficient {\em graph partitioning}, in
which reads contributing to disconnected subgraphs are subdivided into
different files.  This can lead to an approximately 20-fold decrease
in the amount of memory needed for metagenome assembly
\cite{Pell2012}, and may also separate reads into species-specific
bins \cite{Howe2012}.

\subsection*{Digital normalization}

We have implemented a streaming ``lossy compression'' algorithm in
khmer.  This ``digital normalization'' algorithm eliminates redundant
short reads while retaining sufficient information to generate a
contig assembly \cite{Brown2012}.  The algorithm takes advantage of the online
k-mer counting functionaliy in khmer to estimate per-read coverage as
reads are examined; reads can then be accepted as novel or rejected as
redundant. This functions as a kind of error correction, because the
net effect is to decrease not only the total number of reads considered
for assembly, but also decreases the total number of errors considered
by the assembler.

%Protocols

%Pipeline/filter

%API freeze

%\subsection*{Subsection 1}

%\subsection*{Subsection 2}

%\section*{Discussion}

% You may title this section "Methods" or "Models". 
% "Models" is not a valid title for PLoS ONE authors. However, PLoS ONE
% authors may use "Analysis" 
%\section*{Materials and Methods}

% Do NOT remove this, even if you are not including acknowledgments
\section*{Acknowledgments}

khmer development has largely been supported by AFRI Competitive Grant
no. 2010-65205-20361 from the USDA NIFA, and is now funded by the
National Human Genome Research Institute of the National Institutes of
Health under Award Number R01HG007513, both to CTB.

%\section*{References}
% The bibtex filename
\bibliography{khmer-software}

%\section*{Figure Legends}
%\begin{figure}[!ht]
%\begin{center}
%%\includegraphics[width=4in]{figure_name.2.eps}
%\end{center}
%\caption{
%{\bf Bold the first sentence.}  Rest of figure 2  caption.  Caption 
%should be left justified, as specified by the options to the caption 
%package.
%}
%\label{Figure_label}
%\end{figure}


%\section*{Tables}
%\begin{table}[!ht]
%\caption{
%\bf{Table title}}
%\begin{tabular}{|c|c|c|}
%table information
%\end{tabular}
%\begin{flushleft}Table caption
%\end{flushleft}
%\label{tab:label}
% \end{table}

\end{document}

