%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% Welcome to writeLaTeX --- just edit your LaTeX on the left,
% and we'll compile it for you on the right. If you give 
% someone the link to this page, they can edit at the same
% time. See the help menu above for more info. Enjoy!
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
% For more detailed article preparation guidelines, please see:
% http://f1000research.com/author-guidelines and http://f1000research.com/data-preparation

\documentclass[10pt,a4paper,twocolumn]{article}
\usepackage{f1000_styles}

\begin{document}

\title{The khmer software package: enabling efficient nucleotide sequence analysis}
\author[1]{Michael R. Crusoe}
\author[2]{Hussien Alameldin}
\author[3]{Sherine Awad}
\author[4]{Elmar Bucher}
\author[5]{Adam Caldwell}
\author[6]{Reed Cartwright}
\author[7]{Bede Constantinides}
\author[8]{Greg Edvenson}
\author[9]{Scott Fay}
\author[10]{Jake Fenton}
\author[11]{Thomas Fenzl}
\author[12]{Jordan Fish}
\author[13]{Leonor Garcia-Gutierrez}
\author[14]{Phillip Garland}
\author[15]{Jonathan Gluck}
\author[16]{Ivan Gonzalez}
\author[17]{Sarah Guermond}
\author[18]{Jiarong Guo}
\author[19]{Aditi Gupta}
\author[20]{Josh Herr}
\author[21]{Adina Howe}
\author[22]{Alex Hyer}
\author[23]{Andreas Härpfer}
\author[24]{Luiz Irber}
\author[25]{Rhys Kidd}
\author[26]{Heather L. Wiencko}
\author[27]{David Lin}
\author[28]{Justin Lippi}
\author[29]{Tamer Mansour}
\author[30]{Pamela McA\'Nulty}
\author[31]{Eric McDonald}
\author[32]{Jessica Mizzi}
\author[33]{Kevin Murray}
\author[34]{Joshua Nahum}
\author[35]{Kaben Nanlohy}
\author[36]{Lex Nederbragt}
\author[37]{Humberto Ortiz-Zuazaga}
\author[38]{Jeramia Ory}
\author[39]{Jason Pell}
\author[40]{Chuck Pepe-Ranney}
\author[41]{Rodney Picett}
\author[42]{Ryan R. Boyce}
\author[43]{Erich Schwarz}
\author[44]{Camille Scott}
\author[45]{Josiah Seaman}
\author[46]{Scott Sievert}
\author[47]{Jared Simpson}
\author[48]{Connor T. Skennerton}
\author[49]{James Spencer}
\author[50]{Ramakrishnan Rajaram Srinivasan}
\author[51]{Daniel Standage}
\author[52]{James Stapleton}
\author[53]{Susan Steinman}
\author[54]{Joe Stein}
\author[55]{Benjamin Taylor}
\author[56]{Will Trimble}
\author[57]{Michael Wright}
\author[58]{Brian Wyss}
\author[59]{Qingpeng Zhang}
\author[60]{en zyme}
\author[61]{C. Titus Brown}
\affil[1]{mcrusoe@msu.edu, \\ Microbiology and Molecular Genetics, \\ Michigan State University, \\ East Lansing, MI 48824}
\affil[2]{hussien@msu.edu}
\affil[3]{sherine.awad@gmail.com, \\ Microbiology and Molecular Genetics, \\ Michigan State University, \\ East Lansing, MI 48824; \\Population Health and Reproduction, \\University of California, Davis, \\ Davis, CA 95616, USA}
\affil[4]{buchere@ohsu.edu}
\affil[5]{adam.caldwell@gmail.com}
\affil[6]{cartwright@asu.edu}
\affil[7]{greg@edvenson.com, \\Pico Computing, Inc., \\Seattle, WA 98104}
\affil[8]{bedeabc@gmail.com}
\affil[9]{scott.fay@invitae.com}
\affil[10]{bocajnotnef@gmail.com}
\affil[11]{thomas.fenzl@gmx.net}
\affil[12]{jrdn.fish@gmail.com}
\affil[13]{l.garcia-gutierrez@warwick.ac.uk}
\affil[14]{pgarland@gmail.com}
\affil[15]{jonathangluck08854@gmail.com}
\affil[16]{iglpdc@gmail.com}
\affil[17]{sarah.guermond@gmail.com}
\affil[18]{guojiaro@gmail.com}
\affil[19]{aditi9783@gmail.com}
\affil[20]{joshua.r.herr@gmail.com}
\affil[21]{howead@msu.edu, \\Microbiology and Molecular Genetics, \\Michigan State University, \\East Lansing, MI 48824}
\affil[22]{theonehyer@gmail.com}
\affil[23]{ahaerpfer@gmail.com}
\affil[24]{luiz.irber@gmail.com, \\Computer Science and Engineering, \\Michigan State University, \\East Lansing, MI 48824, USA}
\affil[25]{rhyskidd@gmail.com}
\affil[26]{wienckhl@tcd.ie}
\affil[27]{dave@verdematics.com}
\affil[28]{jlippi@gmail.com}
\affil[29]{drtamermansour@gmail.com}
\affil[30]{pamela@addgene.org}
\affil[31]{em@msu.edu}
\affil[32]{mizzijes@msu.edu}
\affil[33]{spam@kdmurray.id.au}
\affil[34]{joshnahum@gmail.com}
\affil[35]{kaben.nanlohy@gmail.com}
\affil[36]{lex.nederbragt@ibv.uio.no}
\affil[37]{humberto.ortiz@upr.edu}
\affil[38]{jeramia.ory@gmail.com}
\affil[39]{jason.pell@gmail.com, \\Computer Science and Engineering, \\Michigan State University, \\East Lansing, MI 48824, USA}
\affil[40]{chuck.peperanney@gmail.com}
\affil[41]{pickett.rodney@gmail.com}
\affil[42]{boycerya@msu.edu}
\affil[43]{ems394@cornell.edu}
\affil[44]{camille.scott.w@gmail.com, \\Computer Science and Engineering, \\Michigan State University, \\East Lansing, MI 48824, USA}
\affil[45]{josiah@dnaskittle.com}
\affil[46]{sieve121@umn.edu}
\affil[47]{js18@sanger.ac.uk}
\affil[48]{c.skennerton@gmail.com}
\affil[49]{james.s.spencer@gmail.com}
\affil[50]{ramrs@nyu.edu}
\affil[51]{daniel.standage@gmail.com}
\affil[52]{jas@msu.edu}
\affil[53]{steinman.tutoring@gmail.com}
\affil[54]{joeaarons@gmail.com}
\affil[55]{taylo886@msu.edu}
\affil[56]{trimble@anl.gov}
\affil[57]{wrigh517@gmail.com}
\affil[58]{wyssbria@msu.edu}
\affil[59]{qingpeng@msu.edu, \\Computer Science and Engineering, \\Michigan State University, \\East Lansing, MI 48824, USA}
\affil[60]{en_zyme@outlook.com}
\affil[61]{zachary.n.russ@gmail.com}
\affil[62]{titus@idyll.org, corresponding author\\Computer Science and Engineering & Microbiology and Molecular Genetics, \\Michigan State University, \\ East Lansing, MI 48824, USA; \\Population Health and Reproduction, \\University of California, Davis, \\Davis, CA 95616, USA}

\maketitle
\thispagestyle{fancy}

Please list all authors that played a significant role in developing the software tool and/or writing the article. Please provide full affiliation information (including full institutional address, ZIP code and e-mail address) for all authors, and identify who is/are the corresponding author(s).

\begin{abstract}

The khmer package is a freely available software library for working efficiently with fixed length DNA words, or k-mers.  khmer provides implementations of a probabilistic k-mer counting data structure, a compressible De Bruijn graph representation, De Bruijn graph
partitioning, and digital normalization.  khmer is implemented in C++ and Python, and is freely available under the BSD license at http://github.com/dib-lab/khmer/.

\end{abstract}
\listoftodos[F1000Research review comments] % Ignore until review stage
\clearpage

\section*{Introduction}

The introduction provides context as to why the software tool was developed and what need it addresses.  It is good scholarly practice to mention previously developed tools that address similar needs, and why the current tool is needed. 

DNA words of a fixed-length k, or ``k-mer'', are a common abstraction
in DNA sequence analysis that enable alignment-free sequence analysis
and comparison. With the advent of second-generation
sequencing and the widespread adoption of De Bruijn graph-based
assemblers, k-mers became even more widely used.  However, the
dramatically increased rate of sequence data generation from the Roche
454 and the Solexa/Illumina sequencers has challenged the basic
data structures and algorithms for k-mer storage and manipulation.
This has led to the development of a wide range of data structures and
algorithms that explore possible improvements to k-mer-based
approaches.

Here we present version 1.4 of the khmer software package, a
high-performance library implementing memory- and time-efficient
algorithms for the manipulation and analysis of short-read data sets.  khmer
contains reference implementations of several approaches, including a
probabilistic k-mer counter based on the CountMin Sketch \cite{Zhang2013}, a
compressible De Bruijn graph representation built on top of Bloom
filters \cite{Pell2012}, and a streaming lossy compression approach for
short-read data sets termed ``digital normalization'' \cite{Brown2012}.

khmer is both research software and a software product for users: it
has been used in the development of novel data structures and
algorithms, and it is also immediately useful for certain kinds of
data analysis (discussed below).  We are continuing to develop research
extensions while maintaining existing functionality.

The khmer software consists of a core library implemented in C++, a
CPython library wrapper implemented in C, and a set of Python
``driver'' scripts that make use of the library to perform various
sequence analysis tasks.  The software is currently developed on
GitHub under github.com/dib-lab/, and it is released under the BSD
License.  There is greater than 80\% statement coverage under
automated tests, measured on both C++ and Python code but primarily
executed at the Python level.

\section*{Methods}
\subsection*{Implementation}
For software tool papers, this section should address how the tool works and any relevant technical details required for implementation of the tool by other developers.  

The core data structures and traversal code are implemented in C++, and
then wrapped for Python in hand-written C code, for a total of
12.2k lines of C/C++ code.  The command-line API and all of the tests
are written in 6.6k lines of Python code.

Documentation is written in reStructuredText, compiled with Sphinx, and hosted on ReadTheDocs.org.

\subsection*{Operation}
This part of the methods should include the minimal system requirements needed to run the software and an overview of the workflow for the tool for users of the tool.

\section*{Features}

khmer has several complementary feature sets, all centered on short-read
manipulation and filtering.

\subsection*{k-mer counting and read trimming}

Using a memory-efficient CountMin Sketch approach, khmer provides an
interface for online counting of k-mers in streams of reads.  The
basic functionality includes calculating the k-mer frequency spectra
in sequence data sets and trimming reads at low-abundance k-mers.
This functionality is explored in \cite{Zhang2013}.

\subsection*{De Bruijn graph representation and exploration}

We have also built a De Bruijn graph representation on top of a Bloom
filter, and implemented this in khmer.  The primary use for this so
far has been to enable memory efficient {\em graph partitioning}, in
which reads contributing to disconnected subgraphs are subdivided into
different files.  This can lead to an approximately 20-fold decrease
in the amount of memory needed for metagenome assembly
\cite{Pell2012}, and may also separate reads into species-specific
bins \cite{Howe2012}.

\subsection*{Digital normalization}

We have implemented a streaming ``lossy compression'' algorithm in
khmer.  This ``digital normalization'' algorithm eliminates redundant
short reads while retaining sufficient information to generate a
contig assembly \cite{Brown2012}.  The algorithm takes advantage of the online
k-mer counting functionaliy in khmer to estimate per-read coverage as
reads are examined; reads can then be accepted as novel or rejected as
redundant. This functions as a kind of error correction, because the
net effect is to decrease not only the total number of reads considered
for assembly, but also decreases the total number of errors considered
by the assembler.


\section*{Results} % Optional - only if novel data or analyses are included
This section is only required if the paper includes novel data or analyses, and should be written as a traditional results section.

\section*{Use Cases} % Optional - only if NO new datasets are included
This section is required if the paper does not include novel data or analyses. 
Examples of input and output files should be provided with some explanatory context.  Any novel or complex variable parameters should also be explained in sufficient detail to allow users to understand and use the tool's functionality.

\section*{Discussion} % Optional - only if novel data or analyses are included
This section is only required if the paper includes novel data or analyses, and should be written in the same style as a traditional discussion section.
Please include a brief discussion of allowances made (if any) for controlling bias or unwanted sources of variability, and the limitations of any novel datasets.


\section*{Conclusions} % Optional - only if novel data or analyses are included
This section is only required if the paper includes novel data or analyses, and should be written as a traditional conclusion.

\section*{Summary} % Optional - only if NO new datasets are included
This section is required if the paper does not include novel data or analyses.  It allows authors to briefly summarize the key points from the article.

\section*{Data availability} % Optional - only if novel data or analyses are included
Please add details of where any datasets that are mentioned in the paper, and that have not have not previously been formally published, can be found.  If previously published datasets are mentioned, these should be cited in the references, as per usual scholarly conventions.

\section*{Software availability}
This section will be generated by the Editorial Office before publication. Authors are asked to provide some initial information to assist the Editorial Office, as detailed below.
\begin{enumerate}
\item URL link to where the software can be downloaded from or used by a non-coder (AUTHOR TO PROVIDE; optional)
khmer.readthedocs.org/en/v1.4/
\item URL link to the author's version control system repository containing the source code
(AUTHOR TO PROVIDE; required)
https://github.com/dib-lab/khmer/tree/v1.4
\item Link to source code as at time of publication ({\textit{F1000Research}} TO GENERATE)
\item Link to archived source code as at time of publication ({\textit{F1000Research}} TO GENERATE)
\item Software license (AUTHOR TO PROVIDE; required)
Copyright (c) 2010-2015, Michigan State University. All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
Neither the name of the Michigan State University nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\end{enumerate}



\section*{Author contributions}
CTB is the primary investigator for the khmer software package. MRC is the lead software developer from July 2013 onwards. Many significant components of khmer have their own paper describing them CITE CITE CITE. The remaining authors each have one or more Git commits in their name.

\section*{Competing interests}
All financial, personal, or professional competing interests for any of the authors that
could be construed to unduly influence the content of the article must be disclosed and will be displayed alongside the article. If there are no relevant competing interests to declare, please add the following: 'No competing interests were disclosed'.

\section*{Grant information}
Please state who funded the work discussed in this article, whether it is your employer, a grant funder etc. Please do not list funding that you have that is not relevant to this
specific piece of research. For each funder, please state the funder’s name, the grant
number where applicable, and the individual to whom the grant was assigned.
If your work was not funded by any grants, please include the line: ‘The author(s)
declared that no grants were involved in supporting this work.’

khmer development has largely been supported by AFRI Competitive Grant
no. 2010-65205-20361 from the USDA NIFA, and is now funded by the
National Human Genome Research Institute of the National Institutes of
Health under Award Number R01HG007513, both to CTB.

\section*{Acknowledgments}
This section should acknowledge anyone who contributed to the research or the
article but who does not qualify as an author based on the criteria provided earlier
(e.g. someone or an organization that provided writing assistance). Please state how
they contributed; authors should obtain permission to acknowledge from all those
mentioned in the Acknowledgments section.

Please do not list grant funding in this section.


\nocite{*}
{\small\bibliographystyle{unsrt}
\bibliography{sample}}


References can be listed in any standard referencing style and should be consistent between references within a given article.

Reference management systems such as Zotero provide options for exporting bibliographies as BibTeX files. BibTex is a bibliographic tool that is used with LaTeX to help organize the user's references and create a bibliography. This template contains an example of such a file, sample.bib, which can be replaced with your own. 


\section*{USING LATEX}
Some examples of commonly used \LaTeX{}  commands and features are listed below, to help you get started.


\subsection*{Sections}

Use section and subsection commands to organize your document. \LaTeX{} handles all the formatting and numbering automatically. Use ref and label commands for cross-references.




% See this guide for more information on BibTeX:
% http://libguides.mit.edu/content.php?pid=55482&sid=406343

% For more author guidance please see:
% http://f1000research.com/author-guidelines


% When all authors are happy with the paper, use the 
% ‘Submit to F1000RESEARCH' button from the Share menu above
% to submit directly to the open life science journal F1000Research.

% Please note that this template results in a draft pre-submission PDF document.
% Articles will be professionally typeset when accepted for publication.

% We hope you find the F1000Research writeLaTeX template useful,
% please let us know if you have any feedback using the help menu above.


\end{document}